File: hot_chocolate_small/dps/entities/Hit.cs
using System;
using System.Collections.Generic;

using HotChocolate;

using x10.hotchoc.dps.Repositories;

namespace x10.hotchoc.dps.Entities {
  // Enums
  public enum UrgencyEnum {
    Low,
    Medium,
    High,
  }


  /// <summary>
  /// A "hit" on a User or Company Entity as reported by LexisNexis
  /// </summary>
  public class Hit : Base {
    // Regular Attributes
    public UrgencyEnum? Urgency { get; set; }
    public HitStatusEnum? Status { get; set; }
    public ReasonForClearanceEnum? ReasonForClearance { get; set; }
    [GraphQLNonNullType]
    public string? Notes { get; set; }

    // To String Representation
    [GraphQLNonNullType]
    public string? ToStringRepresentation {
      get { return "Hit: " + DbidHotChoc; }
      set { /* Needed to make Hot Chocolate happy */ }
    }

    // Associations
    public CompanyEntity? CompanyEntity { get; set; }
    public User? User { get; set; }
    [GraphQLNonNullType]
    public List<DpsAttachment>? Attachments { get; set; }
    [GraphQLNonNullType]
    public List<MatchInfo>? Matches { get; set; }
    [GraphQLNonNullType]
    public List<Shipment>? Shipments { get; set; }
    [GraphQLNonNullType]
    public List<Quote>? Quotes { get; set; }
    [GraphQLNonNullType]
    public List<Booking>? Bookings { get; set; }
    [GraphQLNonNullType]
    public List<DpsMessage>? Messages { get; set; }
    [GraphQLNonNullType]
    public List<OldHit>? OldHits { get; set; }
    [GraphQLNonNullType]
    public WhitelistDuration? WhitelistDays { get; set; }

    public override void EnsureUniqueDbid() {
      base.EnsureUniqueDbid();
      Attachments?.ForEach(x => x.EnsureUniqueDbid());
      Matches?.ForEach(x => x.EnsureUniqueDbid());
      Shipments?.ForEach(x => x.EnsureUniqueDbid());
      Quotes?.ForEach(x => x.EnsureUniqueDbid());
      Bookings?.ForEach(x => x.EnsureUniqueDbid());
      Messages?.ForEach(x => x.EnsureUniqueDbid());
      OldHits?.ForEach(x => x.EnsureUniqueDbid());
    }

    internal override void SetNonOwnedAssociations(IRepository repository) {
      base.SetNonOwnedAssociations(repository);

      int? companyEntity = IdUtils.FromFrontEndId(CompanyEntity?.Id);
      CompanyEntity = companyEntity == null ? null : repository.GetCompanyEntity(companyEntity.Value);

      int? user = IdUtils.FromFrontEndId(User?.Id);
      User = user == null ? null : repository.GetUser(user.Value);

      if (Attachments != null)
        foreach (DpsAttachment attachments in Attachments)
          attachments.SetNonOwnedAssociations(repository);

      if (Matches != null)
        foreach (MatchInfo matches in Matches)
          matches.SetNonOwnedAssociations(repository);

      if (Shipments != null)
        foreach (Shipment shipments in Shipments)
          shipments.SetNonOwnedAssociations(repository);

      if (Quotes != null)
        foreach (Quote quotes in Quotes)
          quotes.SetNonOwnedAssociations(repository);

      if (Bookings != null)
        foreach (Booking bookings in Bookings)
          bookings.SetNonOwnedAssociations(repository);

      if (Messages != null)
        foreach (DpsMessage messages in Messages)
          messages.SetNonOwnedAssociations(repository);

      if (OldHits != null)
        foreach (OldHit oldHits in OldHits)
          oldHits.SetNonOwnedAssociations(repository);

      int? whitelistDays = IdUtils.FromFrontEndId(WhitelistDays?.Id);
      WhitelistDays = whitelistDays == null ? null : repository.GetWhitelistDuration(whitelistDays.Value);
    }
  }
}

File: platforms/flow/react_small_generated/x10_generated/dps/entities/Hit.js
// TEAM: compliance
// @flow

import { v4 as uuid } from 'uuid';

import { addError, type FormError } from 'react_lib/form/FormProvider';
import isBlank from 'react_lib/utils/isBlank';
import toEnum from 'react_lib/utils/toEnum';

import { bookingCalculateErrors, type Booking } from 'dps/entities/Booking';
import { type CompanyEntity } from 'dps/entities/CompanyEntity';
import { dpsAttachmentCalculateErrors, type DpsAttachment } from 'dps/entities/DpsAttachment';
import { dpsMessageCalculateErrors, type DpsMessage } from 'dps/entities/DpsMessage';
import { matchInfoCalculateErrors, type MatchInfo } from 'dps/entities/MatchInfo';
import { oldHitCalculateErrors, type OldHit } from 'dps/entities/OldHit';
import { quoteCalculateErrors, type Quote } from 'dps/entities/Quote';
import { shipmentCalculateErrors, type Shipment } from 'dps/entities/Shipment';
import { type User } from 'dps/entities/User';
import { type WhitelistDuration } from 'dps/entities/WhitelistDuration';
import { type HitStatusEnum, type ReasonForClearanceEnum } from 'dps/sharedEnums';


// Type Definition
export type Hit = {
  +id: string,
  +urgency: ?UrgencyEnum,
  +status: ?HitStatusEnum,
  +reasonForClearance: ?ReasonForClearanceEnum,
  +notes: string,
  +companyEntity: ?CompanyEntity,
  +user: ?User,
  +attachments: $ReadOnlyArray<DpsAttachment>,
  +matches: $ReadOnlyArray<MatchInfo>,
  +shipments: $ReadOnlyArray<Shipment>,
  +quotes: $ReadOnlyArray<Quote>,
  +bookings: $ReadOnlyArray<Booking>,
  +messages: $ReadOnlyArray<DpsMessage>,
  +oldHits: $ReadOnlyArray<OldHit>,
  +whitelistDays: ?WhitelistDuration,
};


// Enums
export const UrgencyEnumPairs = [
  {
    value: 'low',
    label: 'Low',
  },
  {
    value: 'medium',
    label: 'Medium',
  },
  {
    value: 'high',
    label: 'High',
  },
];

export type UrgencyEnum = 'low' | 'medium' | 'high';



// Create Default Function
export function createDefaultHit(): Hit {
  return {
    id: uuid(),
    urgency: null,
    status: 'unresolved',
    // $FlowExpectedError Required field, but no default value
    reasonForClearance: null,
    notes: '',
    companyEntity: null,
    user: null,
    attachments: [],
    matches: [],
    shipments: [],
    quotes: [],
    bookings: [],
    messages: [],
    oldHits: [],
    whitelistDays: null,
  };
}


// Validations
export function hitCalculateErrors(hit: Hit, prefix?: string, inListIndex?: number): $ReadOnlyArray<FormError> {
  const errors = [];
  if (hit == null ) return errors;

  if (isBlank(hit.reasonForClearance))
    addError(errors, prefix, 'Reason For Clearance is required', ['reasonForClearance'], inListIndex);
  if (isBlank(hit.notes))
    addError(errors, prefix, 'Notes is required', ['notes'], inListIndex);
  if (isBlank(hit.whitelistDays))
    addError(errors, prefix, 'Whitelist Days is required', ['whitelistDays'], inListIndex);

  hit.attachments?.forEach((x, ii) => errors.push(...dpsAttachmentCalculateErrors(x, 'attachments', ii)));
  hit.matches?.forEach((x, ii) => errors.push(...matchInfoCalculateErrors(x, 'matches', ii)));
  hit.shipments?.forEach((x, ii) => errors.push(...shipmentCalculateErrors(x, 'shipments', ii)));
  hit.quotes?.forEach((x, ii) => errors.push(...quoteCalculateErrors(x, 'quotes', ii)));
  hit.bookings?.forEach((x, ii) => errors.push(...bookingCalculateErrors(x, 'bookings', ii)));
  hit.messages?.forEach((x, ii) => errors.push(...dpsMessageCalculateErrors(x, 'messages', ii)));
  hit.oldHits?.forEach((x, ii) => errors.push(...oldHitCalculateErrors(x, 'oldHits', ii)));

  if (toEnum(hit?.status) == "unresolved")
    addError(errors, prefix, 'Please select one of the choices above', ['status'], inListIndex);

  return errors;
}

